classes/external/search_packages.php

```diff
@@ -0,0 +1,399 @@
+     */
+    private static function sql_where(string $condition): string {
+        if (trim($condition) === '') {
+            return '';
+        }
+        return "WHERE $condition";
+    }
+
+    /**
+     * This method is called when the service is executed.
+     *
+     * @param string $query
+     * @param array $tags
+     * @param string $category
+     * @param string $sort
+     * @param string $order
+     * @param int $limit
+     * @param int $offset
+     * @return array
+     * @throws moodle_exception
+     */
+    public static function execute(string $query, array $tags, string $category, string $sort, string $order,
+                                   int $limit, int $offset): array {
+        global $DB;
+
+        // Basic parameter validation.
+        $params = self::validate_parameters(self::execute_parameters(), [
+            'query' => $query,
+            'tags' => $tags,
+            'category' => $category,
+            'sort' => $sort,
+            'order' => $order,
+            'limit' => $limit,
+            'offset' => $offset,
+        ]);
+
+        // In addition to the basic parameter validation we also want to validate the values.
+        self::validate_parameter_values($params);
+
+        // Create the 'ORDER BY' sql fragment.
+        $orderbysql = self::create_order_by_sql($params['sort'], $params['order']);
+
+        // Search only in best package translation.
+        [$joinlangssql, $joinlangsparams, $coalescenamesql, $coalescedescsql] = self::create_best_language_sql();
+
+        // Get only packages with specified tags.
+        [$jointagssql, $conditionssql, $jointagsparams] = self::create_tag_filter_sql($params['tags']);
+        $wheretagsql = self::sql_where($conditionssql);
+
+        // Prepare query.
+        [$likesql, $likeparams] = self::create_text_search_sql(['name', 'description'], $params['query']);
+        $wherelikesql = self::sql_where($likesql);
+
+        // Assemble the final sql.
+        // TODO: assemble the sql in a separate function and take the categories into account.
+        $finalsql = "
+            SELECT id, short_name, namespace, author, url, icon, license, name, description
+            FROM (
+                SELECT p.id, p.shortname AS short_name, p.namespace, p.author, p.url, p.icon, p.license,
+                       $coalescenamesql AS name, $coalescedescsql AS description, p.timecreated
+                FROM {qtype_questionpy_package} p
+                $joinlangssql
+                $jointagssql
+                $wheretagsql
+            ) subq
+            $wherelikesql
+            $orderbysql
+        ";
+
+        // Merge every parameter array.
+        $finalparams = array_merge($joinlangsparams, $jointagsparams, $likeparams);
+
+        // Execute the assembled sql query and set the limit and offset.
+        $packagesraw = $DB->get_records_sql($finalsql, $finalparams, $params['limit'] * $params['offset'], $params['limit']);
```

Comment:
`offset` sollte entweder ein absoluter Wert ohne Multiplikation sein oder wir nennen den Parameter in `page` oder so um. Ich tendiere gerade zur Umbenennung.